module/jobqueue/consumer.go

```diff
@@ -0,0 +1,310 @@
+	// for instance, job 3 is the last processed job, job 4 is being processed, job 5 is
+	// done, then we say job 5 is finished, because job 4 hasn't
+	// been processed. Once job 4 is done, both job 4 and 5 will become processed.
+	// maxFinished is the max number of finished jobs to have before stopping working
+	// on any new jobs.
+
+	// State Variables
+	running    bool         // a signal to control whether to process more jobs. Useful for waiting until the workers
+	isChecking *atomic.Bool // allow only one process checking job processable
+	// are ready, and stop when shutting down.
+
+	processedIndex   int64
+	processings      map[int64]*jobStatus // keep track of the status of each on going job
+	processingsIndex map[JobID]int64      // lookup the index of the job, useful when fast forwarding the
+	// `processed` variable
+}
+
+func NewConsumer(
+	log zerolog.Logger,
+	jobs storage.Jobs,
+	progress storage.ConsumerProgress,
+	worker Worker,
+	maxProcessing int64,
+	maxFinished int64) *Consumer {
+	return &Consumer{
+		log: log,
+
+		// store dependency
+		jobs:     jobs,
+		progress: progress,
+		worker:   worker,
+
+		// update config
+		maxProcessing: maxProcessing,
+		maxFinished:   maxFinished,
+
+		// init state variables
+		running:          false,
+		isChecking:       atomic.NewBool(false),
+		processedIndex:   0,
+		processings:      make(map[int64]*jobStatus),
+		processingsIndex: make(map[JobID]int64),
+	}
+}
+
+// Start starts consuming the jobs from the job queue.
+func (c *Consumer) Start() error {
+	c.Lock()
+	defer c.Unlock()
+
+	if c.running {
+		return nil
+	}
+
+	c.running = true
+
+	// on startup, sync with storage for the processed index
+	// to ensure the consistency
+	processedIndex, err := c.progress.ProcessedIndex()
+	if errors.Is(err, storage.ErrNotFound) {
+		processedIndex, err = c.progress.InitProcessedIndex()
+		if err != nil {
+			return fmt.Errorf("could not init processed index: %w", err)
+		}
+		c.log.Info().Int64("processed", processedIndex).
+			Msg("processed index not found, initialized")
+	}
+
+	if err != nil {
+		return fmt.Errorf("could not read processed index: %w", err)
+	}
+
+	c.processedIndex = processedIndex
+
+	c.checkProcessable()
+
+	c.log.Info().Int64("processed", processedIndex).Msg("consumer started")
+	return nil
+}
+
+// Stop stops consuming jobs from the job queue.
+// Note, it won't stop the existing worker from finishing their job
+func (c *Consumer) Stop() {
+	c.Lock()
+	defer c.Unlock()
+
+	c.running = false
+	c.log.Info().Msg("consumer stopped")
+}
+
+// FinishJob let the consumer know a job has been finished, so that consumer will take
+// the next job from the job queue if there are workers available
+func (c *Consumer) FinishJob(jobID JobID) {
+	c.Lock()
+	defer c.Unlock()
+	c.log.Debug().Str("job_id", string(jobID)).Msg("finishing job")
+
+	if c.doneJob(jobID) {
+		c.checkProcessable()
```

Comment:
`Check` will attempt to acquire the lock, calling it from inside of FinishJob (which is before releasing the lock) would cause deadlock.

Calls to both `FinishJob` and `Check` could be concurrent